File: search/binary_search.ex
defmodule Ok do
  def search(list, target) do
    binary_search(list, target, 0, length(list) - 1)
  end

  defp binary_search(list, target, low, high) do
    if low <= high do
      mid = div(low + high, 2)

      case Enum.at(list, mid) do
        ^target -> true
        x when x > target -> binary_search(list, target, low, mid - 1)
        _ -> binary_search(list, target, mid + 1, high)
      end
    else
      false
    end
  end
end

defmodule Optimal do
  def search(list, target), do: binary_search(list, target, 0, length(list) - 1)

  defp binary_search(list, target, low, high) do
    if low <= high do
      mid = div(low + high, 2)

      case Enum.at(list, mid) do
        ^target -> true
        x when x > target -> binary_search(list, target, low, mid - 1)
        _ -> binary_search(list, target, mid + 1, high)
      end
    else
      false
    end
  end
end

File: search/binary-search.go
package main

import (
	"fmt"
	"math"
)

func binarySearch(arr []int, target int) int {
	low := arr[0]
	high := arr[len(arr)-1]

	for low < high {
		mid := math.Floor(float64(low+high) / 2)
		midToInt := int(mid)

		if arr[midToInt] == target {
			return midToInt
		} else if arr[midToInt] < target {
			low = midToInt - 1
		} else {
			high = midToInt + 1
		}
	}
	return -1
}

func main() {
	arr := []int{1, 2, 3, 4, 5, 6, 7, 8, 9}
	n := binarySearch(arr, 4)

	fmt.Println(n)

}
